### DISCUSSION TITLE
Chapter 10. 회복과 병행 제어

### DISCUSSION THREAD

**Discussion by jinny-l** [CodeSquad-2023-BE-Study/DB-Study#14] [완료]
책을 읽으면서 궁금한 점을 남겨주세요.

**Comment by jinny-l** [comment]
p412
> 교착 상태는 처음부터 발생하지 않도록 예방하거나, 발생했을 때 빨리 탐지하여 필요한 조치를 취하는 방법으로 해결한다.

흠... 개발자나 DBA가 예방하지 못한 케이스에서 데드락이 발생하면 서비스 장애일텐데 예방하기 위한 조치가 따로 없으려나요...

**Reply by new-pow** [reply]
데드락이 발생하면 어떻게 해결해야하는가?? 갑자기 궁금해졌어요

**Reply by jinny-l** [reply]
데이터베이스 타임아웃 관련 내용을 아직 잘 모르지만
특정 시간내에 해결이 안되면 타임아웃되도록 설정해야 할 것 같아요

**Reply by jinny-l** [reply]
https://ummmmchicken.tistory.com/entry/%EC%9A%B4%EC%98%81%EC%B2%B4%EC%A0%9C-%EB%8D%B0%EB%93%9C%EB%9D%BD-DeadLock-%EA%B5%90%EC%B0%A9-%EC%83%81%ED%83%9C

**Comment by jinny-l** [comment]
10챕터 내용 어렵네요 ☹️

**Comment by new-pow** [comment]
p381
> 트랜잭션의 내부 문제가 아닌, 하드웨어의 이상이나 소프트웨어의 오류로 트랜잭션의 수행이 중단되독 철회 상태가 된 경우에는 철회된 트랜잭션을 다시 시작한다.
- 이부분은 회복을 한다는 뜻이겠죠?

**Reply by new-pow** [reply]
언제 어떤 방법으로 회복을 하는지 궁금했는데 뒤에 이어서 계속 내용이 나오는 것 같습니다.

**Reply by new-pow** [reply]
- 논리적인 예외의 경우 : 메모리에 있는 것을 토대로 회복 (redo, undo)
- 물리적인 예외의 경우 : 디스크에 있는 것을 토대로 회복 (input + log, bump )

**Reply by new-pow** [reply]
acid
D durability 영구적 저장

**Comment by new-pow** [comment]
p411
- 2단계 로킹 규약에 대해 이해가 잘 안돼요 😢
- T1이 unlock(X) 하기 전에 lock(Y)를 하면서 모든 lock 연산을 실행하고, 그 다음 축소 단계로 들어간 것인가요?

**Reply by jinny-l** [reply]
인터리빙(Interleaving)은 여러 개의 스레드 또는 작업이 동시에 실행될 때, 각 작업이 번갈아 가며 실행되는 현상을 말합니다. 이것은 병렬 처리나 동시성 작업에서 주의해야 할 현상 중 하나입니다.

예를 들어, 두 개의 스레드가 동시에 실행되는 상황에서 A 스레드와 B 스레드가 번갈아 가며 실행되는 것을 인터리빙이라고 합니다. 각 스레드 또는 작업은 일정한 단위로 실행되며, 그 사이에 컨텍스트 스위칭이 발생하여 번갈아 가며 실행됩니다.

인터리빙은 병렬 처리와 동시성을 다룰 때 주의해야 할 점 중 하나입니다. 인터리빙으로 인해 발생하는 경합 상태(Race Condition)와 같은 문제를 예방하고 해결하기 위해 동기화 기법과 스레드 안전성을 고려해야 합니다.

스레드 간 상호작용이나 공유 자원의 접근을 관리하거나, 순서가 중요한 작업에서 인터리빙을 적절히 제어하기 위해 동기화 메커니즘을 사용하거나 락(Lock)을 적용하는 등의 방법을 활용합니다. 이를 통해 스레드 간 경쟁 상태와 인터리빙으로 인한 문제를 최소화하고 안정적인 동시성을 유지할 수 있습니다.

By GPT
